src/solver.rs

```diff
@@ -8,8 +10,67 @@ pub fn query_smt(query: &String) -> bool {
 }
 
 pub fn query_smt_with_solver(query: &String, solver: SolverConfig) -> bool {
-    let query = format!("{}\n{}", query, "(check-sat)");
+    let (output, stderr) = query_smt_internal(&format!("{query}\n(check-sat)\n"), solver);
 
+    match output.as_str() {
+        "sat\n" => true,
+        "unsat\n" => false,
+        _ => panic!(
+            "Invalid output from smt solver.\nQuery: {query}\nStdout: {output}\nStderr: {stderr}",
+        ),
+    }
+}
+
+pub fn query_smt_with_solver_and_eval(
+    query: &String,
+    exprs_to_eval: &Vec<String>,
+    solver: SolverConfig,
+) -> (bool, Vec<BigUint>) {
+    // TODO we are calling the solver twice because we do not know
+    // if the query is sat.
+    if !query_smt_with_solver(query, solver.clone()) {
+        (false, vec![])
+    } else if exprs_to_eval.is_empty() {
+        (true, vec![])
+    } else {
+        let (output, stderr) = query_smt_internal(
+            &format!(
+                "(set-option :produce-models true)\n{query}\n(check-sat)\n(get-value ({}))\n",
+                exprs_to_eval.join(" ")
+            ),
+            solver,
+        );
+        let lines = output.split('\n').collect::<Vec<_>>();
+        if lines.len() < 2 || lines[0] != "sat" {
+            panic!(
+                "SMT solver did not respond with \"sat\".\nQuery: {query}\nStdout: {output}\nStderr: {stderr}",
+            );
+        }
+        // This is a very crude parser.
+        let values = lines[1].to_string();
+        assert!(values.starts_with("(("));
+        assert!(values.ends_with("))"));
+
+        (
+            true,
+            (values[2..values.len() - 2])
+                .split(") (")
+                .map(|pair| {
+                    let v = pair.split(' ').nth(1).unwrap();
+                    if let Some(v) = v.strip_prefix("#b") {
+                        BigUint::from_str_radix(v, 2).unwrap()
+                    } else if let Some(v) = &v.strip_prefix("#x") {
+                        BigUint::from_str_radix(v, 16).unwrap()
+                    } else {
+                        panic!("Invalid value: {v}");
```

Comment:
I guess this will be trigged for models of arrays and functions

Reply:
I'll write some tests.